storage-retrieval/lib/src/client.rs

```diff
@@ -0,0 +1,252 @@
+    InitSwarm(#[from] InitSwarmError),
+    /// This error indicates that the download was timed out.
+    #[error("Download timeout")]
+    DownloadTimeout,
+    /// Error occurred when trying to establish or upgrade an outbound connection.
+    #[error("Dial error: {0}")]
+    Dial(#[from] DialError),
+    /// Error produced by the mater
+    #[error("Mater error: {0}")]
+    Mater(#[from] mater::Error),
+}
+
+/// A client is used to download blocks from the storage provider. Single client
+/// supports getting a single payload.
+pub struct Client {
+    /// Providers of data
+    providers: Vec<Multiaddr>,
+    /// Swarm instance
+    swarm: Swarm<Behaviour<EmptyBlockstore>>,
+    /// The in flight block queries. If empty we know that the client received
+    /// all requested data.
+    queries: HashMap<QueryId, Cid>,
+    /// Blockstore used by the client to store blocks into.
+    blockstore: FileBlockstore,
+    /// Content roots being downloaded.
+    roots: Vec<Cid>,
+    /// Timeout used to cancel the download if not finished before the specified duration.
+    timeout: Option<Duration>,
+}
+
+impl Client {
+    pub async fn new<P>(
+        path: P,
+        providers: Vec<Multiaddr>,
+        roots: Vec<Cid>,
+        timeout: Option<Duration>,
+    ) -> Result<Self, ClientError>
+    where
+        P: AsRef<Path>,
+    {
+        // The p2p node which is created by the client doesn't need a real
+        // blockstore. The reason is that the blockstore is only used by the
+        // node when sharing blocks with other peers.
+        let swarm = new_swarm(Arc::new(EmptyBlockstore))?;
+
+        // Blockstore used to store blocks in. The reason why we separated the
+        // actual blockstore used by the client and the blockstore passed to the
+        // swarm is, because the bitswap behaviour is adding blocks to the
+        // blockstore in asynchronous manner. Because of that we couldn't know
+        // when was the download actually finished.
+        let blockstore = FileBlockstore::new(path, roots.clone()).await?;
+
+        Ok(Self {
+            providers,
+            swarm,
+            queries: HashMap::new(),
+            blockstore,
+            roots,
+            timeout,
+        })
+    }
+
+    /// Start download of some content with a payload cid.
+    pub async fn download(mut self) -> Result<(), ClientError> {
+        // Dial all providers
+        for provider in self.providers.clone() {
+            self.swarm.dial(provider)?;
+        }
+
+        // sleep(Duration::from_secs(1)).await;
+
+        // Start the download by requesting the roots of the trees.
+        self.roots
+            .clone()
+            .into_iter()
+            .for_each(|root| self.request_block(root));
+
+        // Timeout future
+        let mut timeout: BoxFuture<()> = self
+            .timeout
+            .map(|t| sleep(t).boxed())
+            .unwrap_or_else(|| std::future::pending().boxed());
```

Comment:
I'm confused, why not use tokio's timeout?

If it is about doing a global timeout, you can encapsulate the loop or something?